weblate/formats/ttkit.py

```diff
@@ -1490,18 +1490,39 @@ class XWikiPropertiesFormat(PropertiesBaseFormat):
     autoload = ("*.properties",)
     new_translation = "\n"
 
+    # Ensure that not translated units are saved too as missing properties and
+    # comments are preserved as in the original source file.
     def save_content(self, handle):
         current_units = self.all_units
+        store_units = self.store.units
+
+        # We empty the store units since we want to control what we'll serialize
         self.store.units = []
-        # Ensure that not translated units are saved too as missing properties.
+
         for unit in current_units:
-            if unit.unit is None:
-                if not unit.has_content():
-                    unit.unit = unit.mainunit
+            # If the translation unit is missing and the current unit is not
+            # only about comment.
+            if unit.unit is None and unit.has_content():
+
+                # We first check if the unit has not been translated as part of a
+                # new language: in that case the unit is not linked yet.
+                found_store_unit = None
+                for store_unit in store_units:
+                    if unit.context == store_unit.name:
+                        found_store_unit = store_unit
+
+                # If we found a unit for same context not linked, we just link it.
+                if found_store_unit is not None:
+                    unit.unit = found_store_unit
+                # else it's a missing unit: we need to mark it as missing.
                 else:
                     missingunit = self.find_unit(unit.context, unit.source)[0]
```

Comment:
The problem with the tests is caused here. Because of the cleaned up units up there, this doesn't find a matching unit and tries to build a new one. The question is whether it is desired. Another question is why the component is referenced there - at this level only translate-toolkit objects should be involved and nothing from Weblate.

Reply:
> The problem with the tests is caused here. Because of the cleaned up units up there, this doesn't find a matching unit and tries to build a new one. The question is whether it is desired.

So that's actually part of the old code, and yes it's desired: whenever the translation units are not present I want them to be copied from the template store and marked as missing. 
Another way of doing it I thought about would be to generate the entire file with the missing translations when adding a language, the problem is that the method for creating a new language file is a static class method. 

> Another question is why the component is referenced there - at this level only translate-toolkit objects should be involved and nothing from Weblate.

Not sure to follow there.